### ISSUE TITLE
Simple macro to make some variants that have same fields

### ISSUE THREAD

**Issue by loloiccl** [andreaferretti/patty#20]
like

```nim
variant Shape:
  Circle(radius: float)
  Square(diagonal: float)
  Rectangle, Ellipse (width: float, height: float)
```

?

_Originally posted by @Quelklef in https://github.com/andreaferretti/patty/issues/17#issuecomment-417893450_

**Comment by loloiccl** [comment]
This causes invalid indentation.
So, another form of macro is needed if we implement this.

```
   1Hint: this [Processing]
   2Hint: macros [Processing]
   3Hint: sequtils [Processing]
   4this.nim(6, 12) Error: invalid indentation
```

**Comment by Quelklef** [comment]
If it's complaining about invalid indentation, then it's a syntax error and there's nothing that can be done. Macros don't work on the same level as the parser and cannot affect parsing.

**Comment by loloiccl** [comment]
Yes, and I make this issue for discuss what syntax is suit to represent the wanted object type.

**Comment by Quelklef** [comment]
Ah, sorry, I misunderstood.

It would be nice to just allow
```nim
variant Shape:
  ...
  Rectangle(width: float, height: float)
  Ellipse(width: float, height: float)
```
but I imagine that this would be difficult to implement since you would have to detect and combine equivalent variants.

Also, this would come with some surprises. If I then wanted to add
```nim
  Cylinder(radius: float, height: float)
```
this would _seem_ reasonable to a user but can't be done due to Nim restraints on variant types.

**Comment by andreaferretti** [comment]
I don't see why the example with cylinder could not be done. It would share `radius` with `Circle` and `height` with `Rectangle` and `Ellipse`. It would be a little tedious to group all variables correctly, but I think it can be made to work

It would need to generate

```nim
type
  ShapeKind = enum
    Circle, Square, Rectangle, Ellipse, Cylinder
  Shape = object
    case kind: ShapeKind
    of Circle, Cylinder:
      radius: float
    of Square:
      diagonal: float
    of Rectangle, Ellipse:
      width: float
    of Rectangle, Cylinder:
      height: float
```

I did not test that but I believe it's valid Nim

**Comment by andreaferretti** [comment]
[Here](https://github.com/nim-lang/Nim/issues/3629#issuecomment-405520616) is the discussion. @Araq mentions a possible improvement that could be used to make this work, but which is not inside Nim yet
